### PR TITLE
feat: switch storage for better thread safety

### PR DESCRIPTION
**Pull request by qingzhuozhen**
### Summary

- update storage for better thread safety

### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-Kotlin/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  no

### CODE REVIEW

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt [hunk 1]
```diff
@@ -181,23 +245,69 @@ class EventsFileManager(
...
 
     private fun reset() {
         curFile.remove(storageKey)
     }
+
+    private fun handleV1Files() {
+        if (kvs.getLong(storageVersionKey, 1L) > 1L) {
+            return
+        }
+        val unFinishedFiles =
```
Comment (izaaz) [inline core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:290]: Instead of just finishing the files, can we just migrate them over to the new format? That way the read path only needs to worry about the latest format. I'd imagine over time we'd have methods like `handleV1Files`, `handleV2Files` etc and the read path only deals with the latest format

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt [hunk 2]
```diff
@@ -122,50 +146,90 @@ class EventsFileManager(
...
+                val isCurrentVersion = content.endsWith("\n")
+                if (isCurrentVersion) {
+                    // handle current version
+                    val events = JSONArray()
+                    content.split("\n").forEach {
+                        if (it.isNotEmpty()) {
+                            try {
+                                events.put(JSONObject(it))
+                            } catch (e: JSONException) {
+                                logger.error("Failed to parse event: $it")
```
Comment (izaaz) [inline core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:173]: It feels unsafe to drop events in the SDK without us being aware of it. Can we still send these events to the server on a different field and have an alert on the server side to alert us when we run into malformed data on the SDK?

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 1 above)
Comment (justin-fiedler) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:290]: I agree, I would prefer a one time migration to checking the format of each file

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 2 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:173]: Let's discuss it later. Is there a field now could be a good fit?

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 1 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:290]: I guess it is just we prefer the io at migration time or read time. Originally thought maybe we could save some effort as we need to handle the read io anyway, though agreed it could be more cleanerfor read once we migrate completely.

Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt]: Migration updated to run at first. Kept the reading for old version files for now just in case anything happened in the migration process.

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 2 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:178]: Added the basic diagonostircs, feel free to let me know if we would like to put the interface in a different way. Thanks. cc @justin-fiedler

File: core/src/main/java/com/amplitude/core/utilities/HttpClient.kt [hunk 3]
```diff
@@ -111,13 +111,13 @@ internal class HttpClient(
...
     val inputStream: InputStream?,
-    val outputStream: OutputStream?
+    val outputStream: OutputStream?,
 ) : Closeable {
-
     private lateinit var apiKey: String
     private lateinit var clientUploadTime: String
     private lateinit var events: String
     private var minIdLength: Int? = null
+    private var diagnostics: String? = null
```
Comment (izaaz) [inline core/src/main/java/com/amplitude/core/utilities/HttpClient.kt:120]: Can we make the diagnostics class a typed class? So we know what all diagnostics we are capturing from within the SDK.

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt [hunk 4]
```diff
@@ -1,68 +1,93 @@
...
 ) {
-    init {
-        createDirectory(directory)
-    }
-
     private val fileIndexKey = "amplitude.events.file.index.$storageKey"
+    private val storageVersionKey = "amplitude.events.file.version.$storageKey"
+    val filePathSet: MutableSet<String> = Collections.newSetFromMap(ConcurrentHashMap<String, Boolean>())
+    val curFile: MutableMap<String, File> = ConcurrentHashMap<String, File>()
+    private val diagnostics = Diagnostics()
```
Comment (izaaz) [inline core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:31]: Diagnostics feels a little generic. Does it make sense for diagnostics to be scoped to the SDK instead of scoped to the EventsFileManager? I'd imagine we can collect diagnostics from other places too and not just within EventsFileManager. 

If you don't want scope creep and want to just focus on malformed events and just collect that here.

File: core/src/main/java/com/amplitude/core/utilities/HttpClient.kt (hunk 3 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/HttpClient.kt:120]: So you mean we pass the diagnostics object here? I guess in HTTP client we could possibly keep the string, while if we could put it in while amplitude client level

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 4 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:31]: I am still debating about this and would like to gather the thoughts. Put here for the first iteration and reuse the mutex from the EventsFileManager, though I am considering if I should move this out to make it available to all.

Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt]: Migrated this on the amplitude instance level so it could possibly cover more if needed.

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt [hunk 5]
```diff
@@ -74,9 +99,10 @@ class EventsFileManager(
      */
     fun read(): List<String> {
         // we need to filter out .temp file, since it's operating on the writing thread
-        val fileList = directory.listFiles { _, name ->
-            name.contains(storageKey) && !name.endsWith(".tmp")
-        } ?: emptyArray()
+        val fileList =
+            directory.listFiles { _, name ->
+                name.contains(storageKey) && !name.endsWith(".tmp") && !name.endsWith(".properties")
```
Comment (yuhao900914) [inline core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:104]: what's the file end with `.properties`?

File: core/src/main/java/com/amplitude/core/Amplitude.kt [hunk 6]
```diff
@@ -320,12 +355,17 @@ open class Amplitude internal constructor(
      * @return the Amplitude instance
      */
     @JvmOverloads
-    fun setGroup(groupType: String, groupName: String, options: EventOptions? = null): Amplitude {
+    fun setGroup(
```
Comment (justin-fiedler) [inline core/src/main/java/com/amplitude/core/Amplitude.kt:358]: Did the lint settings change? Seems like a lot of style updates

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt [hunk 7]
```diff
@@ -122,50 +146,96 @@ class EventsFileManager(
...
+                    val normalizedContent = "[${content.trimStart('[', ',').trimEnd(']', ',')}]"
+                    try {
+                        val jsonArray = JSONArray(normalizedContent)
+                        return@use jsonArray.toString()
+                    } catch (e: JSONException) {
+                        diagnostics.addMalformedEvent(normalizedContent)
+                        logger.error("Failed to parse events: $normalizedContent, dropping file: $filePath")
+                        this.remove(filePath)
+                        return@use ""
+                    }
```
Comment (justin-fiedler) [inline core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:194]: Is this block still needed after the migration?

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt [hunk 8]
```diff
@@ -181,23 +251,92 @@ class EventsFileManager(
...
-        file.createNewFile()
-        FileOutputStream(file).use {
-            it.write(content.toByteArray())
-            it.flush()
+    private fun writeEventsToSplitFile(
+        events: List<JSONObject>,
+        file: File,
+        append: Boolean = true,
+    ) {
+        val contents = events.joinToString(separator = "\n", postfix = "\n") { it.toString() }
```
Comment (justin-fiedler) [inline core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:284]: Is there any chance of `\n` in the event name or properties?

File: core/src/main/java/com/amplitude/core/utilities/Diagnostics.kt [hunk 9]
```diff
@@ -0,0 +1,47 @@
+package com.amplitude.core.utilities
+
+import java.util.Collections
+
+class Diagnostics() {
```
Comment (justin-fiedler) [inline core/src/main/java/com/amplitude/core/utilities/Diagnostics.kt:5]: Can we make this `internal` for now?

**Review by justin-fiedler — COMMENTED** [review]
@qingzhuozhen overall looks good. See nits/questions. Thanks for adding diagnostics!

File: core/src/main/java/com/amplitude/core/Amplitude.kt (hunk 6 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/Amplitude.kt:358]: Yeah, I wonder if it is for the ktlint version installed is different than before..Would you prefer to revert it back?

File: core/src/main/java/com/amplitude/core/utilities/Diagnostics.kt (hunk 9 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/Diagnostics.kt:5]: That's what I started first, however, after change it to use in the constructor of EventsFileManager and refer that in android module I need to make this public

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 5 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:104]: When it is pure FileStorage, we are using .properties file to store the property. Found this during unit tests so exclude it. In Android this won't happen.

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 7 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:194]: In theory it is not needed, I only left this just in case if there is something wrong in migration we still have a backup. Would you prefer we remove this now or it would be better to remove later?

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 8 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:284]: Great question! I believe that's possible at least for event or property values. In this write piece this should be fine, though I am a bit more worried on the read part. Will test that a bit more and see how that goes. BTW, any suggestions for handling that?

Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt]: @justin-fiedler I tried to use the url encode/decode to handle the line break issue and hopefully that helps, let me know your thoughts.

File: core/src/main/java/com/amplitude/core/Amplitude.kt (hunk 6 above)
Comment (justin-fiedler) [reply core/src/main/java/com/amplitude/core/Amplitude.kt:358]: No preference on the style as long as it stays consistent. Just wanted to check if this was expected.

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 7 above)
Comment (justin-fiedler) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:205]: Sounds good, I'm ok with this as a safety net for now. Can you maybe add a comment that section is not expected to be called and can be removed later based on diagnostics?

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 8 above)
Comment (justin-fiedler) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt]: Thanks. I think the encoding should fix it but am a little concerned about the extra processing for all events. I wonder if there is some other character we can use for the delimiter instead of `\n` that user's are not capable of entering e.g. some special unicode value?

Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt]: @justin-fiedler I am not sure if there is other values we could use. I guess if we use other value, we still need to handle that value is possibly part of the event or property value. We could discuss later in sync about this.

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 7 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt:205]: Sure, I can add a comment for this.

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 8 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt]: Just updated to use the new delimiter as discussed.

File: core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt (hunk 7 above)
Comment (izaaz) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt]: Maybe in the else condition we just send the data as malformed events?

Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt]: @izaaz Just to confirm, would you prefer we return the malformed events or just remove the whole block? In theory this block won't be hit. Though given the edge cases we hit earlier I guess a safety guard won't hurt.

Comment (izaaz) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt]: I mean we send this data as malformed data to the server so that we can recover if we run into bugs like we did earlier. The problem with safe guards is that we'll start adding it for every new version of the file and this function will soon start getting harder to read.

Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt]: @izaaz I updated to return the whole content if it falls into malformed case, just kept the handling at first. So we could reduce the case for server to handle. Won't worry much about the new version changes as we could always refine and clean up this part.

**Review by justin-fiedler — APPROVED** [review]
Look good. Thanks @qingzhuozhen!

**Review by izaaz — APPROVED** [review]
